### PR TITLE
Ifdef out legacy uses of GT_ASG_op

### PR DESCRIPTION
**Pull request by mikedn**
GT_ASG_op nodes are only generated when the legacy backend is used.

### CODE REVIEW

**Comment by mikedn** [comment]
No diffs.

Both ETW and PIN claim a 0.3% reduction in instructions retired (with PGO disabled). Seems too good to be true but there certainly is some code related to ASG_ops that runs even if it's not needed.

**Comment by BruceForstall** [comment]
@dotnet/jit-contrib

File: src/jit/compiler.hpp [hunk 1]
```diff
@@ -1639,12 +1639,13 @@ inline bool GenTree::IsVarAddr() const
...
 {
 #if !defined(_TARGET_64BIT_) && !defined(LEGACY_BACKEND)
-    assert(gtOper == GT_MUL || gtOper == GT_CAST || gtOper == GT_ADD || gtOper == GT_SUB || gtOper == GT_ASG_ADD ||
-           gtOper == GT_ASG_SUB || gtOper == GT_ADD_LO || gtOper == GT_SUB_LO || gtOper == GT_ADD_HI ||
-           gtOper == GT_SUB_HI);
-#else
+    assert(gtOper == GT_MUL || gtOper == GT_CAST || gtOper == GT_ADD || gtOper == GT_SUB || gtOper == GT_ADD_LO ||
+           gtOper == GT_SUB_LO || gtOper == GT_ADD_HI || gtOper == GT_SUB_HI);
+#elif LEGACY_BACKEND
     assert(gtOper == GT_MUL || gtOper == GT_CAST || gtOper == GT_ADD || gtOper == GT_SUB || gtOper == GT_ASG_ADD ||
```
Comment (BruceForstall) [inline src/jit/compiler.hpp]: This should be `#elif defined(LEGACY_BACKEND)`

File: src/jit/lsraarm.cpp [hunk 2]
```diff
@@ -364,8 +364,10 @@ void LinearScan::TreeNodeInfoInit(GenTree* tree)
             break;
 
         case GT_ASG:
+#ifdef LEGACY_BACKEND
         case GT_ASG_ADD:
```
Comment (BruceForstall) [inline src/jit/lsraarm.cpp]: LSRA is never legacy; you can just delete this code

File: src/jit/lsraarm64.cpp [hunk 3]
```diff
@@ -198,8 +198,10 @@ void LinearScan::TreeNodeInfoInit(GenTree* tree)
             break;
 
         case GT_ASG:
+#ifdef LEGACY_BACKEND
```
Comment (BruceForstall) [inline src/jit/lsraarm64.cpp]: LSRA is never legacy; you can just delete this code

File: src/jit/lsraxarch.cpp [hunk 4]
```diff
@@ -294,8 +294,10 @@ void LinearScan::TreeNodeInfoInit(GenTree* tree)
             break;
 
         case GT_ASG:
+#ifdef LEGACY_BACKEND
```
Comment (BruceForstall) [inline src/jit/lsraxarch.cpp]: LSRA is never legacy; you can just delete this code

File: src/jit/morph.cpp [hunk 5]
```diff
@@ -16526,8 +16540,7 @@ void Compiler::fgMorphStmts(BasicBlock* block, bool* mult, bool* lnot, bool* loa
             continue;
         }
 
-#if OPT_MULT_ADDSUB
-
+#if defined(OPT_MULT_ADDSUB) && defined(LEGACY_BACKEND)
         /* Note whether we have two or more +=/-= operators in a row */
```
Comment (BruceForstall) [inline src/jit/morph.cpp]: This should be `#if OPT_MULT_ADDSUB && defined(LEGACY_BACKEND)`

`OPT_MULT_ADDSUB` is 0 or 1; `LEGACY_BACKEND` is defined or not defined.

File: src/jit/morph.cpp [hunk 6]
```diff
@@ -16686,7 +16700,7 @@ void Compiler::fgMorphBlocks()
 
         fgMorphStmts(block, &mult, &lnot, &loadw);
 
-#if OPT_MULT_ADDSUB
+#if defined(OPT_MULT_ADDSUB) && defined(LEGACY_BACKEND)
```
Comment (BruceForstall) [inline src/jit/morph.cpp]: same here

File: src/jit/morph.cpp (hunk 5 above)
Comment (mikedn) [reply src/jit/morph.cpp]: Oops, missed that. It looks like OPT_MULT_ADDSUB is pointless it's always 1. The whole thing reduces to `ifdef LEGACY_BACKEND`.

File: src/jit/emitarm.cpp [hunk 7]
```diff
@@ -7889,7 +7889,11 @@ regNumber emitter::emitInsTernary(instruction ins, emitAttr attr, GenTree* dst,
             jumpKind                = isUnsignedOverflow ? EJ_lo : EJ_vs;
             if (jumpKind == EJ_lo)
             {
-                if ((dst->OperGet() != GT_SUB) && (dst->OperGet() != GT_ASG_SUB) && (dst->OperGet() != GT_SUB_HI))
+                if ((dst->OperGet() != GT_SUB) &&
+#ifdef LEGACY_BACKEND
+                    (dst->OperGet() != GT_ASG_SUB) &&
+#endif
+                    (dst->OperGet() != GT_SUB_HI))
```
[the remark this answers is not in the document]
Comment (mikedn) [reply src/jit/emitarm.cpp]: I'm not sure adding `OperIsAddOp` & co. is a good idea. They may end up being used in cases where they're not needed (before LIR there's no `GT_XXX_LO`/`HI`) or used incorrectly (for example in `genCodeForBinary` we try to generate inc/dec from `GT_ADD`, that should never be done for `GT_ADD_LO`/`HI`).

In the case of `gtOverflow` we should just add something like `OperMayOverflow` and use that in both `gtOverflow` and `gtOverflowEx`. Then we may consider adding `GTK_OVERFLOW`, though that may turn out to be slower than just checking for a few opers.

**Comment by mikedn** [comment]
I pushed a commit that deals with `gtOverflow` and `gtOverflowEx`. Note that `gtOverflow`'s assert was allowing `GT_ADD/SUB_LO` while `gtOverflowEx`'s condition did not. These opers are not supposed to overflow, it's just that some decomposition code was attempting to check for overflow after changing the oper from `GT_ADD/SUB` to `GT_ADD/SUB_LO`.

In case anyone wonders, code like `return ((gtOper == GT_ADD) || (gtOper == GT_SUB) || (gtOper == GT_MUL) || (gtOper == GT_CAST)` generates
```asm
00007FFEC3F15274 8D 41 D1             lea         eax,[rcx-2Fh]  
00007FFEC3F15277 3C 02                cmp         al,2  
00007FFEC3F15279 76 08                jbe         GenTree::OperMayOverflow+0Fh (07FFEC3F15283h)  
00007FFEC3F1527B 80 F9 1B             cmp         cl,1Bh  
00007FFEC3F1527E 74 03                je          GenTree::OperMayOverflow+0Fh (07FFEC3F15283h)  
```
It might or might not be faster than relying on OperKind that involves a memory load, hard to say. Anyway, in terms of throughput it might be faster to have dedicated opers for overflow operations instead of using flags...

**Comment by mikedn** [comment]
> Unfortunately, it increases #ifdef density in the meantime :-(

Yes, and I really don't like ifdefs. At least now there's less dead code to read (for example, I was looking at `RangeCheck` and was wondering when `GT_ASG_ADD` appears... never)

I think I have address all feedback. Needs squashing (AFAIK it can be done in the GitHub UI during merge?)

**Comment by mikedn** [comment]
And speaking of `RangeCheck`, it looks like some code can be simplified a bit now that there's only one assignment oper. There's no need to store a pointer to the assignment node in `Location`, a pointer to the assignment's source operand should do. But that's a more involved change that doesn't fit here.

File: src/jit/rangecheck.cpp [hunk 8]
```diff
@@ -1034,17 +1044,22 @@ bool RangeCheck::DoesVarDefOverflow(BasicBlock* block, GenTreePtr stmt, GenTreeP
...
 
+#ifdef LEGACY_BACKEND
         case GT_ASG_ADD:
             // For GT_ASG_ADD, op2 is use, op1 is also use since we side table for defs in useasg case.
             return DoesBinOpOverflow(loc->block, loc->stmt, asg->gtGetOp1(), asg->gtGetOp2(), path);
+#endif
 
         default:
+#ifndef LEGACY_BACKEND
+            unreached();
```
Comment (briansull) [inline src/jit/rangecheck.cpp]: Note that this causes a desktop build break due to unreachable code warning

File: src/jit/rangecheck.cpp [hunk 9]
```diff
@@ -875,14 +880,19 @@ Range RangeCheck::ComputeRangeForLocalDef(
...
             // If the operator of the definition is +=, then compute the range of the operands of +.
             // Note that gtGetOp1 will return op1 to be the lhs; in the formulation of ssa, we have
             // a side table for defs and the lhs of a += is considered to be a use for SSA numbering.
             return ComputeRangeForBinOp(loc->block, loc->stmt, asg->gtGetOp1(), asg->gtGetOp2(), GT_ADD, path,
                                         monotonic DEBUGARG(indent));
+#endif
 
         default:
+#ifndef LEGACY_BACKEND
+            unreached();
```
Comment (briansull) [inline src/jit/rangecheck.cpp]: Note that this causes a desktop build break due to unreachable code warning

File: src/jit/rangecheck.cpp [hunk 10]
```diff
@@ -384,16 +384,21 @@ bool RangeCheck::IsMonotonicallyIncreasing(GenTreePtr expr, SearchPath* path)
...
                 return IsMonotonicallyIncreasing(asg->gtGetOp2(), path);
 
+#ifdef LEGACY_BACKEND
             case GT_ASG_ADD:
                 return IsBinOpMonotonicallyIncreasing(asg->gtGetOp1(), asg->gtGetOp2(), GT_ADD, path);
+#endif
 
             default:
+#ifndef LEGACY_BACKEND
+                unreached();
```
Comment (briansull) [inline src/jit/rangecheck.cpp]: Note that this causes a desktop build break due to unreachable code warning

File: src/jit/rangecheck.cpp (hunk 8 above)
Comment (mikedn) [reply src/jit/rangecheck.cpp]: Hmm, what's different between the desktop build and the CoreCLR build? Compilation options?

Comment (BruceForstall) [reply src/jit/rangecheck.cpp]: Possibly. And different toolsets.

Comment (mikedn) [reply src/jit/rangecheck.cpp]: It looks like one difference is that the desktop build uses /W4 while the CoreCLR build uses /W3

**Pull request merged**